Make the SDK's representation of bounds data internal

shsms opened the pull request:
This make the `inclusion_bounds` and `exclusion_bounds` fields of the
battery pool's `Report` class private, and instead introduces a
`bounds` property as a single set of bounds that users can use.

This PR also introduces the `Report.adjust_to_bounds` method which can
be used in case more granularity is needed on what exact power values
can be proposed to the battery pool.

shsms committed:
De-clutter assignment statements

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Document what bounds overlapping with exclusion bounds look like

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Expose `PowerManagingActor`'s `Report` type through the `BatteryPool`

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Allow `_clamp_to_bounds` to clamp power values to both sides

In cases where there's an exclusion zone within the usable bounds, the
given power can be clamped to both ends of the exclusion zone.

This function now returns both sides, rather than just the one that's
closest to the given power.

This can also be used in the bounds interface, exposed through the
`Report` class.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Move bounds methods to a separate module

This allows them to be reused in the battery pool, without leading to
circular imports.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add method for adjusting power values to battery pool's reports

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Expose only outer usable bounds

With this change, the SDK will just expose one set of outer bounds,
even if it has exclusion zones within.  The representation of bounds
data in the SDK is no longer exposed.

When users need additional control over what power values they
propose, they can use the `adjust_to_bounds` methods.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms commented:
Here too,  there's a line already in the release notes providing sufficient context:  https://github.com/frequenz-floss/frequenz-sdk-python/blob/7da41bf58f53892fcb354f0df4370e2f7120ed85/RELEASE_NOTES.md?plain=1#L26

So not adding anything more.

shsms commented on src/frequenz/sdk/actor/_power_managing/_base_classes.py:
```diff
@@ -72,6 +73,80 @@ class Report:
+        """
+        return self._inclusion_bounds
+
+    def adjust_to_bounds(self, power: Power) -> tuple[Power | None, Power | None]:
```
Still need to figure out what the right return type is.  They are not bounds, but the closest available power values that can be proposed to the battery pool.

That said,  this whole method can be safely ignored and people can just rely on the PowerManager, so there's no need to make this more clear.

There's also an example below in this method's docstring, that shows how it can be used with a `match` statement.  Maybe that's all that's needed.

llucax replied:
Why publishing it if you don't have to use it then?

If you want to keep it, about the type, maybe `PowerRange`?

![PowerRangers](https://media.nbcboston.com/2023/01/GettyImages-819239982.jpg?quality=85&strip=all&crop=0px%2C161px%2C2752px%2C1548px&resize=1200%2C675)

shsms replied:
Except it is not a `Range` either,  they are the two best possible options.  

Most users shouldn't have to worry about this method, but it  might be useful in one specific situation:  If an actor sets bounds that fall entirely within the exclusion bounds range, they might want to be able to choose which side of the exclusion bounds they want to go to.

shsms replied:
I'm open to removing it and adding it if there seems to be a need for such flexibility among users.  Especially in cases where there's only a single actor running, they'll be able to get more control by using this, if they want.

shsms replied:
I guess this is one of the problems with trying to stabilize an API that has never been used in production before.

Maybe we just prefix with `_` and call it experimental for now?  Or maybe prefix with `experimental_`, so automatic documentation works?

shsms replied:
Didn't do anything, leave it as it is.  The function returns a pair of powers, which don't have two separate names, just two options, so I think a tuple makes sense.

llucax replied:
I'm all for leaving it with `_`, so if someone needs it, they can use it already, and we don't need to remove it now and add it later. If people really start to use it and it works, we make it public. I really like this approach as "the best of both worlds".

shsms replied:
is there anyway to override the doc generator to show docs for this function even if it is `_`?

llucax replied:
No, but that's the point, right? That it is hidden, but if someone asks for the functionality we can say "use this"?

shsms replied:
This method's documentation provides a lot of context about the behaviour of the bounds, so I'd like it to be visible,  but I'll add a line saying that the method is experimental.

and it looks like it is possible:  https://mkdocstrings.github.io/python/usage/configuration/members/

I will test.

llucax replied:
In the API that's not really possible, as all of that is auto-generated and we don't have a way to override how we render a particular module or symbol. You could pull that documentation into some other doc in `docs/`. What I don't like about making it public is that autocompletion might suggest it and users might not look at the docs in detail, and after you are using it, there no indication that you are using something experimental.

With `_` you don't get auto-completion and the code is marked with `# pylint: disable=private-access` (or whatever is called), which clearly shows something is fishy, and it is also clear that private interfaces are not stable.

Can you put that documentation somewhere else? Maybe in the class, saying "There is an experimental `_adjust_to_bounds()` that does this and that"?

shsms replied:
Actually I'd like to take some more time to think about this,  so I'd like to do this in the next PR.

The next PR still deals with these parts - it makes this class internal, and generic, removing all mention of batteries,  and there will be a `Protocol` type (as I see it now), that would expose the methods, but with specific documentation and corresponding to the pool.

So we won't forget about this if we don't do it now.

shsms replied:
> What I don't like about making it public is that autocompletion might suggest it and users might not look at the docs in detail, and after you are using it, there no indication that you are using something experimental.

I mean, this whole thing is one big experimental feature, not just this function.  We're basically changing all of these interfaces and trying to release them before we've had a chance to test them.

llucax replied:
Fair enough!

matthias-wende-frequenz commented on src/frequenz/sdk/actor/_power_managing/_base_classes.py:
```diff
@@ -49,14 +50,14 @@ class Report:
     """The currently set power for the batteries."""
 
-    inclusion_bounds: timeseries.Bounds[Power] | None
+    _inclusion_bounds: timeseries.Bounds[Power] | None
```
I'd suggest not to use this terminology any longer, but if you decide to do so, it needs documentation.

shsms replied:
We'll be forced to change it when we upgrade to the new microgrid API.  Until then, the internal repr just reflects the API version.  Is why I made them internal anyway, so people don't have to deal with these terms anymore.

matthias-wende-frequenz commented on src/frequenz/sdk/actor/_power_managing/_base_classes.py:
```diff
@@ -72,6 +73,80 @@ class Report:
 
+    @property
+    def bounds(self) -> timeseries.Bounds[Power] | None:
+        """The bounds for the batteries.
```
not only for batteries, right?

shsms replied:
oh true.  I guess we need separate `Report` classes, one for `BatteryPool`, `PVPool`, etc., so they can have precise documentation.  I think that's the only public interface class, so probably not so bad.  

And actually can just be done by inheriting from the PowerManager's report I think.

shsms replied:
unfortunately the word "Battery" is everywhere in the power manager implementation.  I will make a separate PR after this to change that.

llucax commented on src/frequenz/sdk/actor/_power_managing/_bounds_methods.py:
```diff
@@ -0,0 +1,143 @@
+    When only the upper bound overlaps with exclusion bounds, the usable range is
+    between the lower bound and the lower exclusion bound, like below.
+
+      ===lb+++++++ex----ub-------ex===
```
To be honest, this sort of confuses me even more, because here you talk about usable range but the function talks about the exclusion zone.

Suggestion of a more explicit alternative:

> ```
>               lower                        upper
>                 .------ exclusion zone ------.
> -----|✔✔✔✔✔✔✔✔✔✔✔✔|xxxxxxxxxxxxxxx|----------|----
>      `-- usable --'-- exclusion --´
>      |                 overlap    |
>      |                            |
>    lower                        upper
>    bound                        bound
>                       (inside the exclusion zone)
> ```
> 
> Resulting in `(False, True)` because only the upper bound is inside the exclusion zone.

For me with this there is no need to do the other variants, seems clear enough with one example.

shsms replied:
That looks awesome, did you make that by hand?

shsms replied:
done

llucax replied:
> That looks awesome, did you make that by hand?

Yes :blush:

llucax replied:
I would love for GitHub/mkdocs to eventually support [aafigure](https://aafigure.readthedocs.io/en/latest/shortintro.html) (ASCII-art to image conversion).

* Rendered examples: https://aafigure.readthedocs.io/en/latest/examples.html
* Source: https://github.com/aafigure/aafigure/blob/master/documentation/examples.rst

llucax commented on src/frequenz/sdk/actor/_power_managing/_matryoshka.py:
```diff
@@ -217,12 +221,23 @@ def _calc_target_power(
                     exclusion_bounds,
-                )
+                ):
+                    case (None, power) | (power, None) if power:
+                        target_power = power
+                    case (power_low, power_high) if power_low and power_high:
```
:exploding_head: I didn't remember the `if` syntax.

shsms replied:
yes, the whole thing is very similar to the `pcase` in emacs lisp, and the `match` from rust.

llucax replied:
And [D's template constraints](https://dlang.org/spec/template.html#template_constraints) :laughing:

shsms replied:
wow, wouldn't have guessed D also has functional pattern matching.

llucax commented on src/frequenz/sdk/actor/_power_managing/_bounds_methods.py:
What about just `_bounds.py`? Methods sounds weird, specially when they are free functions, and the module name shouldn't really tell if you store classes, or functions or methods, but what do they do.

shsms replied:
true, done.

llucax commented on src/frequenz/sdk/actor/_power_managing/_bounds_methods.py:
```diff
@@ -0,0 +1,143 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Methods for checking and clamping bounds and power values to exclusion bounds."""
```
```suggestion
"""Utilities dealing with bounds, like checking and clamping, and power values to exclusion bounds."""
```

shsms replied:
It is actually clamping of (bounds and power values) to exclusion bounds.  So just changed to: 

> Utilities for checking and clamping bounds and power values to exclusion bounds.

llucax commented on src/frequenz/sdk/actor/_power_managing/_bounds_methods.py:
```diff
@@ -0,0 +1,143 @@
+from ...timeseries import Bounds, Power
+
+
+def _check_exclusion_bounds_overlap(
```
I would make the symbols public now (no underscore), as they are meant to be imported from other modules. Having the module *private* (starting with `_`) should be enough to hide it from the end user interface.

shsms replied:
done

llucax commented on src/frequenz/sdk/actor/_power_managing/_base_classes.py:
```diff
@@ -49,14 +50,14 @@ class Report:
     """The currently set power for the batteries."""
 
-    inclusion_bounds: timeseries.Bounds[Power] | None
+    _inclusion_bounds: timeseries.Bounds[Power] | None
     """The available inclusion bounds for the batteries, for the actor's priority.
 
     These bounds are adjusted to any restrictions placed by actors with higher
     priorities.
     """
 
-    exclusion_bounds: timeseries.Bounds[Power] | None
+    _exclusion_bounds: timeseries.Bounds[Power] | None
```
Minor, but maybe you want to move these to the end of the attributes, so all *public attributes* are listed first.

shsms replied:
done

shsms committed:
Apply suggestions from code review

- Rename `_bounds_methods.py` -> `_bounds.py`.
- Make methods in `_bounds.py` not private.
- Move private attribute definitions in `Report` class to the end.
- Improve docstring example for `check_exclusion_bounds_overlap`
  method.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

llucax commented on src/frequenz/sdk/actor/_power_managing/_bounds.py:
```diff
@@ -0,0 +1,143 @@
+
+    Example:
+
+                       lower                        upper
+                          .----- exclusion zone -----.
+        -----|✓✓✓✓✓✓✓✓✓✓✓✓|xxxxxxxxxxxxxxx|----------|----
+             `-- usable --'-- exclusion --´
+             |                 overlap    |
+             |                            |
+           lower                        upper
+           bound                        bound
+                              (inside the exclusion zone)
```
This will not be rendered for now as it is private, but if at some point we make it public, it will render incorrectly, so I suggest:

~~~suggestion
        ```
                       lower                        upper
                          .----- exclusion zone -----.
        -----|✓✓✓✓✓✓✓✓✓✓✓✓|xxxxxxxxxxxxxxx|----------|----
             `-- usable --'-- exclusion --´
             |                 overlap    |
             |                            |
           lower                        upper
           bound                        bound
                              (inside the exclusion zone)
        ```
~~~

llucax reviewed (commented):
This one small comment and the pending decision on if we make `adjust_to_bounds()` private or not.

shsms committed:
Apply documentation suggestion from code review

Co-authored-by: Leandro Lucarella <luca@llucax.com>
Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

llucax commented:
LGTM, leaving @matthias-wende-frequenz to approve as he also has some pending comments.
